### PR TITLE
Make project settings reflect command-line port option

### LINKED ISSUE #708: Uploading a project with HTTP that has a different port throws an exception
This is caused by the server interrupting the thread that's loading the project when the port changes.

Stacktrace:

```java
edu.wpi.grip.core.exception.GripServerException: Could not stop Jetty server
	at edu.wpi.grip.core.http.GripServer.stop(GripServer.java:175)
	at edu.wpi.grip.core.http.GripServer.setPort(GripServer.java:216)
	at edu.wpi.grip.core.http.GripServer.settingsChanged(GripServer.java:234)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.google.common.eventbus.Subscriber.invokeSubscriberMethod(Subscriber.java:95)
	at com.google.common.eventbus.Subscriber$SynchronizedSubscriber.invokeSubscriberMethod(Subscriber.java:154)
	at com.google.common.eventbus.Subscriber$1.run(Subscriber.java:80)
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:456)
	at com.google.common.eventbus.Subscriber.dispatchEvent(Subscriber.java:76)
	at com.google.common.eventbus.Dispatcher$PerThreadQueuedDispatcher.dispatch(Dispatcher.java:119)
	at com.google.common.eventbus.EventBus.post(EventBus.java:215)
	at edu.wpi.grip.core.serialization.ProjectSettingsConverter.unmarshal(ProjectSettingsConverter.java:33)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshallField(AbstractReflectionConverter.java:503)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.doUnmarshal(AbstractReflectionConverter.java:429)
	at com.thoughtworks.xstream.converters.reflection.AbstractReflectionConverter.unmarshal(AbstractReflectionConverter.java:281)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convert(TreeUnmarshaller.java:72)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:66)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.convertAnother(TreeUnmarshaller.java:50)
	at com.thoughtworks.xstream.core.TreeUnmarshaller.start(TreeUnmarshaller.java:134)
	at com.thoughtworks.xstream.core.AbstractTreeMarshallingStrategy.unmarshal(AbstractTreeMarshallingStrategy.java:32)
	at com.thoughtworks.xstream.XStream.unmarshal(XStream.java:1230)
	at com.thoughtworks.xstream.XStream.unmarshal(XStream.java:1214)
	at com.thoughtworks.xstream.XStream.fromXML(XStream.java:1085)
	at edu.wpi.grip.core.serialization.Project.open(Project.java:115)
	at edu.wpi.grip.core.serialization.Project.open(Project.java:108)
	at edu.wpi.grip.core.http.HttpPipelineSwitcher.handleIfPassed(HttpPipelineSwitcher.java:46)
	at edu.wpi.grip.core.http.PedanticHandler.handle(PedanticHandler.java:53)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:119)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
	at org.eclipse.jetty.server.Server.handle(Server.java:518)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:308)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:244)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95)
	at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceAndRun(ExecuteProduceConsume.java:246)
	at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:156)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:654)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:572)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.InterruptedException
	at java.lang.Object.wait(Native Method)
	at java.lang.Thread.join(Thread.java:1257)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.doStop(QueuedThreadPool.java:161)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.stop(AbstractLifeCycle.java:89)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.stop(ContainerLifeCycle.java:143)
	at org.eclipse.jetty.util.component.ContainerLifeCycle.doStop(ContainerLifeCycle.java:161)
	at org.eclipse.jetty.server.handler.AbstractHandler.doStop(AbstractHandler.java:73)
	at org.eclipse.jetty.server.Server.doStop(Server.java:476)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.stop(AbstractLifeCycle.java:89)
	at edu.wpi.grip.core.http.GripServer.stop(GripServer.java:173)
	... 44 more
```

### PR DESCRIPTION
**Pull request by SamCarlberg**
No longer serializes the server port. Use the command-line option or the settings dialog to set it. Fixes #708

Enable projects to be uploaded via HTTP in UI mode

### CODE REVIEW

File: core/src/main/java/edu/wpi/grip/core/Main.java [hunk 1]
```diff
@@ -83,7 +88,9 @@ public void start(String[] args) throws IOException, InterruptedException {
         } else {
           // Valid port; set it (Note: this doesn't check to see if the port is available)
           logger.info("Running server on port " + port);
-          gripServer.setPort(port);
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/Main.java]: Can you move the branch validation into GripServer.

File: core/src/main/java/edu/wpi/grip/core/http/HttpPipelineSwitcher.java [hunk 2]
```diff
@@ -41,24 +38,8 @@ protected void handleIfPassed(String target,
...
-        response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
-        sendTextContent(response,
-            String.format("GRIP is not in the correct mode: should be HEADLESS, but is %s", mode),
-            CONTENT_TYPE_PLAIN_TEXT);
-        baseRequest.setHandled(true);
-        break;
-    }
+    project.open(new String(IOUtils.toByteArray(request.getInputStream()), "UTF-8"));
+    response.setStatus(HttpServletResponse.SC_CREATED);
+    baseRequest.setHandled(true);
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/http/HttpPipelineSwitcher.java]: Sure

File: core/src/main/java/edu/wpi/grip/core/settings/ProjectSettings.java [hunk 3]
```diff
@@ -43,9 +43,10 @@
   private String deployJvmOptions = "-Xmx50m -XX:-OmitStackTraceInFastThrow "
       + "-XX:+HeapDumpOnOutOfMemoryError";
 
+  // Transient because save files shouldn't be responsible for knowing the server port
   @Setting(label = "Internal server port",
       description = "The port that the internal server should run on.")
-  private int serverPort = 8080;
+  private transient int serverPort = 8080;
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/settings/ProjectSettings.java]: Override on upload?

File: core/src/test/java/edu/wpi/grip/core/http/HttpPipelineSwitcherTest.java [hunk 4]
```diff
@@ -79,7 +78,7 @@ public void open(String projectXml) {
         didRun[0] = true;
       }
     };
-    pipelineSwitcher = new HttpPipelineSwitcher(new ContextStore(), project, GripMode.HEADLESS);
+    pipelineSwitcher = new HttpPipelineSwitcher(new ContextStore(), project);
```
Comment (JLLeitschuh) [inline core/src/test/java/edu/wpi/grip/core/http/HttpPipelineSwitcherTest.java]: Move this constructor into a helper method in the test so you don't have to change all usages of it

File: ui/src/main/java/edu/wpi/grip/ui/Main.java [hunk 5]
```diff
@@ -158,7 +162,9 @@ public void start(Stage stage) throws IOException {
         } else {
           // Valid port; set it (Note: this doesn't check to see if the port is available)
           logger.info("Running server on port " + port);
-          server.setPort(port);
+          ProjectSettings settings = settingsProvider.getProjectSettings().clone();
```
Comment (JLLeitschuh) [inline ui/src/main/java/edu/wpi/grip/ui/Main.java]: Is this clone done somewhere else? Should this be in a common location?

**Comment by JLLeitschuh** [comment]
Solution, putting a divider for properties that are not saved.

**Comment by SamCarlberg** [comment]
Closed in favor of #724

**Pull request closed without merging**